pkg/storage/unified/resource/search_queue.go

```diff
@@ -0,0 +1,145 @@
+package resource
+
+import (
+	"context"
+	"sync"
+	"time"
+)
+
+// indexQueueProcessor manages queue-based operations for a specific index
+// It is responsible for ingesting events for a single index
+// It will batch events and send them to the index in a single bulk request
+type indexQueueProcessor struct {
+	index     ResourceIndex
+	nsr       NamespacedResource
+	queue     chan *WrittenEvent
+	batchSize int
+	builder   DocumentBuilder
+
+	resChan chan *IndexEvent // Channel to send results to the caller
+
+	mu      sync.Mutex
+	running bool
+}
+
+type IndexEvent struct {
+	WrittenEvent      *WrittenEvent
+	Action            IndexAction
+	IndexableDocument *IndexableDocument // empty for delete actions
+	Timestamp         time.Time
+	Latency           time.Duration
+	Err               error
+}
+
+// NewIndexQueueProcessor creates a new IndexQueueProcessor for the given index
+func NewIndexQueueProcessor(index ResourceIndex, nsr NamespacedResource, batchSize int, builder DocumentBuilder, resChan chan *IndexEvent) *indexQueueProcessor {
+	return &indexQueueProcessor{
+		index:     index,
+		nsr:       nsr,
+		queue:     make(chan *WrittenEvent, 1000), // Buffer size of 1000 events
+		batchSize: batchSize,
+		builder:   builder,
+		resChan:   resChan,
+		running:   false,
+	}
+}
+
+// Add adds an event to the queue and ensures the background processor is running
+func (b *indexQueueProcessor) Add(evt *WrittenEvent) {
+	b.queue <- evt
+
+	// Start the processor if it's not already running
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	if !b.running {
+		b.running = true
+		b.startProcessor()
+	}
+}
```

Comment:
The key difference is here. Instead of processing events one by one, we add the event to a queue and start a processor that will process the queue in small batches until it's empty.